src/params.ts

```diff
@@ -0,0 +1,106 @@
+import { InternalError } from "./errors";
+import { FileCache } from "./file-cache";
+
+export enum LoadingStatus {
+    NotStarted = 0,
+    InProgress,
+    Completed,
+    Failed
+}
+
+export class SnarkParams {
+    url: string;
+    expectedHash: string | undefined;
+    params: any | undefined;
+    loadingPromise: Promise<any> | undefined;
+    loadingStatus: LoadingStatus;
+
+    public constructor(url: string, expectedHash: string | undefined) {
+        this.loadingStatus = LoadingStatus.NotStarted;
+        this.url = url;
+        this.expectedHash = expectedHash;
+    }
+
+    public async get(wasm: any): Promise<any> {
+        if (this.loadingStatus == LoadingStatus.Completed) {
+            return this.params;
+        }
+
+        if (this.needToLoad()) {
+            this.load(wasm);
+        }
+
+        return await this.loadingPromise;
```

Comment:
```suggestion
        if (this.needToLoad()) {
            this.load(wasm);
            
            return await this.loadingPromise;
        }
        
        return this.params;
```

Do not diffuse state machine logic between different functions

Reply:
If params are loading now we still want to await the promise.

Reply:
I mean if this.loadingStatus == LoadingStatus.InProgress

Reply:
It's all just about the code clarity, the common logic is correct anyway

Reply:
Agreed, it looks simpler this way.

Reply:
Done: https://github.com/zkBob/zkbob-client-js/pull/80/commits/e70548515e8601111c819bba8ada9ddf6176328c